File: tests/spec/runners/python/comparator.py
"""Result comparison logic for .gtest spec tests.

Mirrors the assertion helpers in ``grafeo-spec-tests/src/lib.rs`` so that the
Python runner validates results identically to the Rust runner.
"""

from __future__ import annotations

import hashlib
import math
from typing import Any, Dict, List, Optional


# ---------------------------------------------------------------------------
# Value to canonical string
# ---------------------------------------------------------------------------


def value_to_string(val: Any) -> str:
    """Convert a Python value (as returned by the grafeo bindings) to its
    canonical string representation for comparison.

    This must match the Rust ``value_to_string`` in
    ``grafeo-spec-tests/src/lib.rs``.
    """
    if val is None:
        return "null"
    if isinstance(val, bool):
        return "true" if val else "false"
    if isinstance(val, int):
        return str(val)
    if isinstance(val, float):
        if math.isnan(val):
            return "NaN"
        if math.isinf(val):
            return "Infinity" if val > 0 else "-Infinity"
        # Rust formats floats without trailing zeros but keeps at least one
        # decimal for non-integers.  Python's str() is close enough for the
        # values typically seen in .gtest files.
        s = str(val)
        return s
    if isinstance(val, list):
        inner = ", ".join(value_to_string(v) for v in val)
        return f"[{inner}]"
    if isinstance(val, dict):
        entries = sorted(f"{k}: {value_to_string(v)}" for k, v in val.items())
        return "{" + ", ".join(entries) + "}"
    if isinstance(val, bytes):
        return f"bytes[{len(val)}]"
    # datetime, date, time: use str() which gives ISO format
    return str(val)


# ---------------------------------------------------------------------------
# Result -> rows of strings
# ---------------------------------------------------------------------------


def result_to_rows(result, columns: Optional[List[str]] = None) -> List[List[str]]:
    """Convert a grafeo QueryResult into rows of canonical strings.

    Each row is obtained by iterating the result (which yields dicts keyed by
    column name), then converting each value via ``value_to_string``.

    If *columns* is provided, values are extracted in that order.  Otherwise
    the column order from ``result.columns`` is used.
    """
    cols = columns or list(result.columns)
    rows: List[List[str]] = []

    # result[i] returns a dict for row i
    for i in range(len(result)):
        row_dict: Dict[str, Any] = result[i]
        row: List[str] = []
        for col in cols:
            row.append(value_to_string(row_dict.get(col)))
        rows.append(row)

    return rows


# ---------------------------------------------------------------------------
# Assertions
# ---------------------------------------------------------------------------


def assert_rows_sorted(
    result,
    expected: List[List[str]],
    columns: Optional[List[str]] = None,
) -> None:
    """Assert that result rows match *expected* after sorting both sides."""
    actual = result_to_rows(result, columns)
    actual_sorted = sorted(actual)
    expected_sorted = sorted(expected)

    assert len(actual_sorted) == len(expected_sorted), (
        f"Row count mismatch: got {len(actual_sorted)} rows, "
        f"expected {len(expected_sorted)}\n"
        f"Actual:   {actual_sorted}\n"
        f"Expected: {expected_sorted}"
    )

    for i, (act, exp) in enumerate(zip(actual_sorted, expected_sorted)):
        assert len(act) == len(exp), (
            f"Column count mismatch at sorted row {i}: "
            f"got {len(act)} cols, expected {len(exp)}\n"
            f"Actual row:   {act}\n"
            f"Expected row: {exp}"
        )
        for j, (a, e) in enumerate(zip(act, exp)):
            assert a == e, (
                f"Mismatch at sorted row {i}, col {j}: "
                f"got '{a}', expected '{e}'\n"
                f"Full actual row:   {act}\n"
                f"Full expected row: {exp}"
            )


def assert_rows_ordered(
    result,
    expected: List[List[str]],
    columns: Optional[List[str]] = None,
) -> None:
    """Assert that result rows match *expected* in exact order."""
    actual = result_to_rows(result, columns)

    assert len(actual) == len(expected), (
        f"Row count mismatch: got {len(actual)} rows, "
        f"expected {len(expected)}\n"
        f"Actual:   {actual}\n"
        f"Expected: {expected}"
    )

    for i, (act, exp) in enumerate(zip(actual, expected)):
        assert len(act) == len(exp), (
            f"Column count mismatch at row {i}: "
            f"got {len(act)} cols, expected {len(exp)}\n"
            f"Actual row:   {act}\n"
            f"Expected row: {exp}"
        )
        for j, (a, e) in enumerate(zip(act, exp)):
            assert a == e, (
                f"Mismatch at row {i}, col {j}: "
                f"got '{a}', expected '{e}'\n"
                f"Full actual row:   {act}\n"
                f"Full expected row: {exp}"
            )


def assert_count(result, expected_count: int) -> None:
    """Assert that the result contains exactly *expected_count* rows."""
    actual = len(result)
    assert actual == expected_count, (
        f"Row count mismatch: got {actual}, expected {expected_count}"
    )


def assert_empty(result) -> None:
    """Assert that the result contains zero rows."""
    actual = len(result)
    assert actual == 0, f"Expected empty result, got {actual} row(s)"


def assert_error(exc: Exception, expected_substr: str) -> None:
    """Assert that the exception message contains *expected_substr*."""
    msg = str(exc)
    assert expected_substr in msg, f"Error '{msg}' does not contain '{expected_substr}'"


def assert_hash(
    result,
    expected_hash: str,
    columns: Optional[List[str]] = None,
) -> None:
    """Assert that the MD5 hash of sorted, pipe-delimited rows matches.

    Mirrors ``assert_hash`` in the Rust runner.
    """
    rows = result_to_rows(result, columns)
    rows.sort()

    hasher = hashlib.md5()
    for row in rows:
        hasher.update("|".join(row).encode())
        hasher.update(b"\n")

    actual_hash = hasher.hexdigest()
    assert actual_hash == expected_hash, (
        f"Hash mismatch: got '{actual_hash}', expected '{expected_hash}'\nRows: {rows}"
    )

File: tests/spec/runners/node/comparator.mjs
/**
 * Result comparison logic for .gtest spec tests.
 *
 * Mirrors the assertion helpers in grafeo-spec-tests/src/lib.rs so that
 * the Node.js runner validates results identically to the Rust runner.
 */

/**
 * Convert a JS value to its canonical string for comparison.
 * Must match Rust's value_to_string in lib.rs.
 */
export function valueToString(val) {
  if (val === null || val === undefined) return 'null'
  if (typeof val === 'boolean') return val ? 'true' : 'false'
  if (typeof val === 'bigint') return val.toString()
  if (typeof val === 'number') {
    if (!isFinite(val)) return val > 0 ? 'Infinity' : '-Infinity'
    if (isNaN(val)) return 'NaN'
    // Rust: format!("{}", 15.0_f64) -> "15" (no trailing .0)
    if (Number.isInteger(val)) return val.toString()
    return val.toString()
  }
  if (Array.isArray(val)) {
    const inner = val.map(valueToString).join(', ')
    return `[${inner}]`
  }
  if (typeof val === 'object' && val !== null) {
    // Date-like objects
    if (val instanceof Date) return val.toISOString()
    // Plain object (map)
    const entries = Object.entries(val)
      .map(([k, v]) => `${k}: ${valueToString(v)}`)
      .sort()
    return `{${entries.join(', ')}}`
  }
  return String(val)
}

/**
 * Convert a GrafeoDB QueryResult to rows of canonical strings.
 * @param {object} result - QueryResult from db.execute()
 * @returns {string[][]}
 */
export function resultToRows(result) {
  const columns = result.columns
  const rows = []
  const arr = result.toArray()
  for (const row of arr) {
    const r = []
    for (const col of columns) {
      r.push(valueToString(row[col]))
    }
    rows.push(r)
  }
  return rows
}

/**
 * Assert rows match after sorting both sides.
 */
export function assertRowsSorted(result, expected) {
  const actual = resultToRows(result)
  const sortedActual = [...actual].sort((a, b) => a.join('|').localeCompare(b.join('|')))
  const sortedExpected = [...expected].sort((a, b) => a.join('|').localeCompare(b.join('|')))

  if (sortedActual.length !== sortedExpected.length) {
    throw new Error(
      `Row count mismatch: got ${sortedActual.length}, expected ${sortedExpected.length}\n` +
      `Actual: ${JSON.stringify(sortedActual)}\nExpected: ${JSON.stringify(sortedExpected)}`
    )
  }
  for (let i = 0; i < sortedActual.length; i++) {
    for (let j = 0; j < sortedActual[i].length; j++) {
      if (sortedActual[i][j] !== sortedExpected[i][j]) {
        throw new Error(
          `Mismatch at sorted row ${i}, col ${j}: got '${sortedActual[i][j]}', expected '${sortedExpected[i][j]}'\n` +
          `Actual row: ${JSON.stringify(sortedActual[i])}\nExpected row: ${JSON.stringify(sortedExpected[i])}`
        )
      }
    }
  }
}

/**
 * Assert rows match in exact order.
 */
export function assertRowsOrdered(result, expected) {
  const actual = resultToRows(result)
  if (actual.length !== expected.length) {
    throw new Error(
      `Row count mismatch: got ${actual.length}, expected ${expected.length}\n` +
      `Actual: ${JSON.stringify(actual)}\nExpected: ${JSON.stringify(expected)}`
    )
  }
  for (let i = 0; i < actual.length; i++) {
    for (let j = 0; j < actual[i].length; j++) {
      if (actual[i][j] !== expected[i][j]) {
        throw new Error(
          `Mismatch at row ${i}, col ${j}: got '${actual[i][j]}', expected '${expected[i][j]}'\n` +
          `Actual row: ${JSON.stringify(actual[i])}\nExpected row: ${JSON.stringify(expected[i])}`
        )
      }
    }
  }
}
